dali/kernels/scratch_copy_impl.h

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SCRATCH_COPY_IMPL_H_
+#define DALI_KERNELS_SCRATCH_COPY_IMPL_H_
+
+#include <cuda_runtime.h>
+#include <algorithm>
+#include <tuple>
+#include <memory>
+#include "dali/core/traits.h"
+#include "dali/kernels/context.h"
+
+namespace dali {
+namespace kernels {
+
+namespace detail {
+
+inline void copy_to_buffer(char *buffer, const size_t *offsets) {}
+
+/// @brief Copy contents of collections `{ c, tail... }` to pointers stored in `ptrs`.
+template <typename Collection, typename... Collections>
+void copy_to_buffer(char *buffer,
+                  const size_t *offsets,
+                  const Collection &c,
+                  const Collections &... tail) {
+  using T = std::remove_cv_t<element_t<Collection>>;
+  std::copy(dali::begin(c), dali::end(c), reinterpret_cast<T*>(buffer + offsets[0]));
+  copy_to_buffer(buffer, offsets+1, tail...);
+}
+
+inline void GetCollectionOffsets(size_t base, size_t *offsets) { *offsets = base; }
+
+/// @brief Allocate storage for contiguous elements from collections `{ c, tail... }`
```

Comment:
I think this description is not aligned with function body.

Reply:
Fixed.